[Docs] Proposal: Concept exercise style guide.

aldraco opened the pull request:
This PR proposes a simple style guide specifically for writing Concept Exercises (with an expectation that it will be added to in the ensuing discussion.)

### Rationale:
We are planning on recruiting the wider community to write Concept Exercises across many tracks. 

While current maintainers might be familiar with Exercism project conventions regarding styles, formatting, and voice, new contributors would not be. A style guide would ensure that (a) contributors know what's expected of them, and (b) maintainers have a consistent point of reference for feedback when needed. 

Since open source pull requests often span weeks of back and forth over implementation detail, a centralized style guide allows us _keep the focus of that conversation on the most valuable content of the exercises_, and not on style change requests. 

This will both make the iterative development and review process more efficient, and make the contributing experience more seamless for newcomers.

**Other Notes**

Note: I have not taken much care yet with the tone/voice of this document, but tried to make it as direct as possible for the sake of discussion here. 

The initial list was generated by collecting comments from existing Concept exercise pull requests that concerned style change requests.

__Not Implemented__
 - a supplemental concept exercise style guide for a particular language


Thanks for any and all feedback. 🎉

iHiD commented on docs/maintainers/concept-exercise-style-guide.md:
````diff
@@ -0,0 +1,56 @@
+[indirect-reference]: https://example.com/link-to-page
+```
+
+Links should be inline where possible, i.e. as `<a href="">text</a>` tags instead of bare URLs. This is accomplished using the Markdown bracketed notation, as exemplified above.
````
It took me a couple of reads to understand this bit, maybe a "do this (example) not this (example)"?

ErikSchierboom replied:
I also found it a bit hard to understand. Could you explain what type of mistake you want to prevent people from making?

aldraco replied:
Yeah, this isn't very clear. Will clean it up. I was primarily referring to not putting links behind text, i.e. leaving the whole URL in the text. An example would help a lot.

ErikSchierboom replied:
Okay great!

iHiD commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+
+## Code
+
+Where possible and relevant, format code first as if it was being typed into a terminal, or if not applicable to your language, as if in a code file itself. Please reference the docs for your language, as they may have more information.
```
I think this would specifically be for examples of code that **we** will run. ie if we show an example of a function definition, I wouldn't want this, but if we were giving an example of the input/output of tests, then I would agree with this.

ErikSchierboom replied:
Agreed. So there are two cases for examples:

1. Code samples to illustrate how something works in a language (usually defined in the `introduction.md` file). These don't need any terminal prompt and such, as that will probably only distract from the actual code.
1. Code samples to show what a piece of code (usually a function/method) returns (usually defined in the `instructions.md` file). These examples would indeed benefit from the terminal prompt bit.

iHiD commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+
+## Miscellaneous
+
+ - Use dashed-style UUIDs in `config.json` files. e.g. `1234-5678-8901`
```
I suggest we link to https://www.uuidgenerator.net/version4

iHiD commented:
I think having this is really great. I'd also like to actually add a bit on tone of voice too.

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+# Formatting and Style Guide for Concept Exercise Files
```
Could you add a link to this new document to the [_sidebar.md document](https://github.com/exercism/v3/blob/master/docs/maintainers/_sidebar.md)?

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+# Formatting and Style Guide for Concept Exercise Files
+
+This document records the current styleguide applicable to Concept exercise Markdown files, which refer to the files contained in `.docs` and `.meta`. Please follow them when implementing a Concept Exercise; maintainers should also reference the Style Guide when reviewing pull requests. When the relevant language also includes a style guide, those preferences will take precedence over the ones he ...
```
```suggestion
This document records the current style guide applicable to Concept exercise Markdown files, which refer to the files contained in an exercise's `.docs` and `.meta` directories. Please follow them when implementing a Concept Exercise; maintainers should also reference the Style Guide when reviewing pull requests. When the relevant language also includes a style guide, those preferences will take precedence over the ones here.
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+ - Use level 2 headers, e.g. `##`
+ - Do not use top-level headers, e.g. `#`
+ - Do not use leading whitespace in the file.
+
```
```suggestion
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+
+If a document is empty, i.e. there is no `after.md` content, then do not check one in.
+
+
```
```suggestion
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+
+## Links
+
+Exercism prefers "indirect reference" links. They are defined at the bottom of the Markdown file and mapped to a reference slug, and referenced by that slug in the text.
```
I think they're called [reference links](https://spec.commonmark.org/0.29/#reference-link). Maybe we can refer to the official docs:

```suggestion
Exercism prefers [reference links](https://spec.commonmark.org/0.29/#reference-link), which are defined at the bottom of the Markdown file, mapped to a reference slug, and referenced by that slug in the text.
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
````diff
@@ -0,0 +1,56 @@
+
+Example:
+
+```
````
I like this example!

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
````diff
@@ -0,0 +1,56 @@
+>>> extract_message("[INFO] Logline message")
+```
+
+
````
```suggestion
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,56 @@
+
+Please consult each language's docs folder for more information on the preferred style conventions for that language.
+
+
```
```suggestion
```

iHiD commented:
> @iHiD any other things you'd like to include on the tone?

Let's do that as a seperate PR so e don't block this.

ErikSchierboom commented:
@aldraco This PR has a merge conflict. Could you look into fixing it, then we can try to merge this PR.

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
````diff
@@ -0,0 +1,56 @@
+
+For example, Python has the REPL (read-eval-print loop), which allows a programmer to type code directly into a terminal, so we prefer formatting example code as if it was being typed into the REPL, with a leading `>>>`:
+
+```
````
```suggestion
```python
```

ErikSchierboom commented:
@aldraco Let me know if you plan on working on this, otherwise I can continue with this awesome work you've already done!

iHiD commented:
@ErikSchierboom I reference this in my latest video so it's probably something we should get onto master this week :)

aldraco commented:
@ErikSchierboom I can pick back up with this PR tonight -- I've had a lot less free time lately, sorry to let this slip as a result. Thanks for the detailed review, and glad it seems like it will be helpful.

I'll check back in later tonight.

aldraco commented:
@iHiD @ErikSchierboom Updated, and tagging for re-review --
 * addressed notes from above
 * fixed merge conflict
 * added a sidebar link from within the maintainers docs folder
 * left out anything on tone/voice for now, as suggested above

Thanks again for the review and y'all's patience. :fireworks:

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
```diff
@@ -0,0 +1,81 @@
+
+## Empty Documents
+
+If a document is empty, i.e. there is no `after.md` content, then do not check one in.
```
The `after.md` file must always be there:

```suggestion
If a document is empty, i.e. there is no `source.md` content, then do not check one in.
```

ErikSchierboom commented on docs/maintainers/concept-exercise-style-guide.md:
````diff
@@ -0,0 +1,81 @@
+
+In other cases, it makes more sense to leave code formatted as if it was in a `.py` file. Here, the student benefits by being presented with something that mimics how they would write the code themselves.
+
+```
````
```suggestion
```python
```

ErikSchierboom commented:
Thanks a lot @aldraco!
